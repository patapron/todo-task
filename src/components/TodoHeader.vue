<template>
  <header class="header">
    <h1>todos</h1>
    <input id="toggle-all" class="toggle-all" type="checkbox" v-model="isAll" />
    <label for="toggle-all"></label>
    <input
      class="new-todo"
      placeholder="New task, press Enter to save"
      autofocus
      @keydown.enter="createTask"
      v-model="task"
    />
  </header>
</template>

<script>
export default {
  props: ["list"],
  data() {
    return {
      task: "",
    };
  },
  methods: {
    /**
     * create new task emit and erro message
     */
    createTask() {
      if (this.task.trim().length > 0) {
        this.$emit("createTaskEmit", this.task);
        this.task = "";
      } else {
        alert("Name field cannot be empty");
      }
    },
  },
  computed: {
    /**
     * check box functions
     */
    isAll: {
      set(checked) {
        this.list.forEach((obj) => (obj.completed = checked));
      },
      get() {
        return (
          this.list.length !== 0 && this.list.every((obj) => obj.completed === true)
        );
      },
    },
  },
};
</script>
