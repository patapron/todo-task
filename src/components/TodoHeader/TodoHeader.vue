<template>
  <header class="header">
    <h1>todos</h1>
    <input
      id="toggle-all"
      class="toggle-all"
      type="checkbox"
      @click="changeToCompleted"
    />
    <label for="toggle-all"></label>
    <input
      :class="{ error: error }"
      class="new-todo"
      placeholder="New task, press Enter to save"
      autofocus
      @keydown.enter="createTask()"
      v-model="taskName"
      @keyup="checkEmpty"
    />
  </header>
</template>

<script lang="ts">
import "./TodoHeader.css";
import { Task } from "@/typings/types";
import Vue, { PropType } from "vue";

export default Vue.extend({
  props: {
    list: {
      type: Array,
      default: () => [] as Task[],
    },
  },
  data() {
    return {
      taskName: "",
      error: false,
    };
  },
  methods: {
    checkEmpty() {
      this.error = this.taskName.trim().length == 0;
    },

    /**
     * create new task emit and erro message
     */
    createTask() {
      if (this.taskName.trim().length > 0) {
        this.$emit("createTaskEmit", this.taskName);
        setTimeout(() => {
          this.taskName = "";
          this.error = false;
        }, 100);
      }
    },
    changeToCompleted() {
      this.$emit("changeToCompletedEmit", this.taskName);
    },
  },
});
</script>
