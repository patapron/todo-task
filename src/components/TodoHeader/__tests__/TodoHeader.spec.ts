/* eslint-disable prettier/prettier */
import TodoHeader from "@/components/TodoHeader/TodoHeader.vue";
import { shallowMount } from "@vue/test-utils";

describe("TodoHeader", () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = shallowMount(TodoHeader, {
      mocks: {
        $t: () => "Test text",
        $can: () => true,
      },
    });
  });

  it("renders correctly", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("createTask", () => {
    wrapper.vm.createTask();
    expect(wrapper.emitted("createTaskEmit")).toBeFalsy();
    wrapper.vm.taskName = "test";
    wrapper.vm.createTask();
    expect(wrapper.emitted("createTaskEmit")).toBeTruthy();
  });
});
