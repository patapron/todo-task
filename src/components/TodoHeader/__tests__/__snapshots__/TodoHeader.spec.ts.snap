// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoHeader renders correctly 1`] = `
<header class="header">
  <h1>todos</h1> <input id="toggle-all" type="checkbox" class="toggle-all"> <label for="toggle-all"></label> <input placeholder="New task, press Enter to save" autofocus="autofocus" class="new-todo">
</header>
`;
